README.md
# az2tf (Python3 version) - now supports Terraform v0.12 

Work in progress - please report any issues you find.

This utility 'Azure to Terraform' (az2tf) 
reads an Azure Subscription and generates all the required terraform configuration files (.tf) from each of the composite Azure Resource Groups
It also imports the terraform state using a

"terraform import ...." command

And finally runs a 

"terraform plan ."  command 

There should hopefully be no subsequent additions or deletions reported by the terraform plan command as all the approriate terraform configuration files will have have automatically been created.

## Requirements & Prerequisites
+ The tool is written for the bash shell script & Python2 and has been tested on macOS
+ Azure cli2 **version 2.0.75 or higher** needs to be installed and you need a login with at least "Read" priviledges
+ terraform **version v0.12.16 or higher** needs to be installed
+ Python **version 3.6.1 or higher**

### May also be required
+ pip install requests
+ pip install adal


## Quickstart guide to using the tool

Running the tool in your local shell (bash) required these steps:
1. Unzip or clone this git repo into an empty directory
2. login to the Azure cli2  (az login)
3. run the tool 


## Usage Guide

### The First Run
To generate the terraform files for a subscription and stop after a "terraform validate":
```
./az2tf.sh -s <Subscription ID> -v yes
```

The above will either show :
```
terraform validate
Success! The configuration is valid.
```

Or there may be some kind of python error. (as trying to test everyone's Azure combinations in advance isn't possible)

**If you happen to find one of these errors please open an issue here and paste in the error and it will get fixed.**

Once the validation is ok you can use the tool in anger to not only generate the terraform files (-v yes) but also import the resources and perform a terraform plan (see below)

---

<br>

To generate the terraform files for an entire Azure subscription, import the resourcs and perform a terraform plan:
```
./az2tf.sh -s <Subscription ID>
```

If your resources are in Azure US Government:
```
./az2tf.sh -c AzureUSGovernment -s <Subscription ID>
```

To include Azure Subscription Policies and RBAC controls and assignments:
```
./az2tf.sh [-c <Cloud Name>] -s <Subscription ID> -p yes
```

To generate the terraform files for a specific Resource Group in a subscription:
```
./az2tf.sh [-c <Cloud Name>] -s <Subscription ID> -g <Resource Group>
```

To include the secrets from a Key Vault in terraform files (secrets will be in plain text):
```
./az2tf.sh [-c <Cloud Name>] -s <Subscription ID> -g <Resource Group> -x yes
```

To filter the terraform resource type: (eg: just availability sets)
```
./az2tf.sh [-c <Cloud Name>] -s <Subscription ID> -g <Resource Group> -r azurerm_availability_set
```
To filter the terraform resource type: (eg: just availability sets) and fast forward - ie. build up resources one after another.:
```
./az2tf.sh -s <Subscription ID> -g <Resource Group> -r azurerm_rsource_group
./az2tf.sh -s <Subscription ID> -g <Resource Group> -r azurerm_availability_set -f yes
./az2tf.sh -s <Subscription ID> -g <Resource Group> -r azurerm_public_ip -f yes
```

To use the fast forward option correctly you'll need a good understanding of terraform resource dependancies to ensure you avoid any depenacy errors.

<br>

Be patient - lots of output is given as az2tf:

+ Loops for each provider through your resource groups &
+ Creates the requited *.tf configuration files in the "generated" directory
+ Performs the necessary 'terraform import' commands
+ And finally runs a 'terraform plan'



## Supported Resource Types

The following terraform resource types are supported by this tool at this time:

Base Resources
* azurerm_resource_group 

Authorization Resources
* azurerm_role_definition (subscription level)
* azurerm_role_assignment (subscription level)
* azurerm_user_assigned_identity

Active Directory Resources

App Service (Web Apps) Resources
* azurerm_app_service
* azurerm_app_service_plan
* azurerm_app_service_slot
* azurerm_function_app

Automation Resources
* azurerm_automation_account

Compute Resources
* azurerm_availability_set
* azurerm_image
* azurerm_managed_disk 
* azurerm_shared_image_gallery
* azurerm_shared_image
* azurerm_shared_image_version
* azurerm_snapshot 
* azurerm_virtual_machine 
* azurerm_virtual_machine_extension 
* azurerm_virtual_machine_scale_set

Container Resources
* azurerm_container_registry 
* azurerm_kubernetes_cluster

CosmosDB (DocumentDB) Resources
* azurerm_cosmosdb_account 

Database Resources
* azurerm_sql_database
* azurerm_sql_server

Databricks Resources
* azurerm_databricks_resources (not available for China Azure today)

Key Vault Resources
* azurerm_key_vault 
* azurerm_key_vault_secret

Load Balancer Resources
* azurerm_lb  
* azurerm_lb_backend_address_pool 
* azurerm_lb_rule 
* azure_nat_rule 
* azurerm_lb_probe
* azure_nat_pool 

Logic App Resources
* azurerm_logic_app_workflow

Management Resources
* azurerm_management_lock 

Messaging Resources
* azurerm_eventhub
* azurerm_eventhub_namespace
* zurerm_eventhub_namespace_authorization_rule

* azurerm_servicebus_namespace 
* azurerm_servicebus_queue 

Monitoring Resources
* azurerm_autoscale_setting  (not available for China Azure today)

Network Resources
* azurerm_application_gateway 
* azurerm_application_security_group 
* azurerm_express_route_circuit 
* azurerm_express_route_circuit_authorization 
* azurerm_express_route_circuit_peering 
* azurerm_local_network_gateway
* azurerm_network_interface  
* azurerm_network_security_group 
* azurerm_network_watcher
* azurerm_public_ip  
* azurerm_route_table 
* azurerm_subnet 
* azurerm_subnet_network_security_group_association 
* azurerm_subnet_route_table_group_association 
* azurerm_traffic_manager_endpoint 
* azurerm_traffic_manager_profile 
* azurerm_virtual_network 
* azurerm_virtual_network_gateway
* azurerm_virtual_network_gateway_connection
* azurerm_virtual_network_peering 
  
Policy Resources
* azurerm_policy_definition (custom poicies only)
* azurerm_policy_set_definition (custom poicies only)
* azurerm_policy_assignment

OMS Resources
* azurerm_log_analytics_solution 
* azurerm_log_analytics_workspace 

Recovery Services
* azurerm_recovery_services_vault 

Storage Resources
* azurerm_storage_account 

## Other Azure Clouds

Global AzureCloud is used by default. To set a specific regional cloud, use `-c <Cloud Name>`

The following are acceptable values:
* AzureCloud
* AzureChinaCloud
* AzureGermanCloud
* AzureUSGovernment

## Planned Additions

+ PaaS databases and other missing providers (feel free to contribute !)
+ ongoing better AKS support as AKS evolves
+ Other terraform providers as terraform supports

## Using the Azure runbook code

Within the runbook directory you'll file a file az2tf-runbook.py

You can paste this code into an Azure python2 runbook, your automation account must have an Azure ARM Runas credential setup.

When runs it will leave you subscriptions terraform and terraform import caommands in the runbook output window

*Unfortunately this only works for small subscriptions as Azure currently has a limit on how much output it will show in a runbook output window. If your output window is blank az2tf has probably worked (see the All Logs tab) but the amount of output has exceeded Azure's limit.*

*To workaround this you can download the output aftert he job has run with powershell see*
https://docs.microsoft.com/azure/automation/automation-runbook-output-and-messages#runbook-output.



If you find any output in the Exceptions tab please open an issue here and report it.


## Known problems

### Speed

It can take a lot of time to loop around everything in large subscriptions, in particular the importing of the resources.

### KeyVault:

Can fail if your login/SPN doesn't have acccess to the KeyVault

### Virtual machines:
These attributes always get reported in terraform plan set to false by default - may need to manually override

+ delete_data_disks_on_termination:           "" => "false"
+ delete_os_disk_on_termination:              "" => "false"


### Storage Account

Can fail if your login/SPN doesn't have acccess to the KeyVault used for encryption.
Can also fail if resource locks are in place

### OMS

If solutions have '[' & ']' in their names they will be ignored.


### ExpressRoute

No support for MS peering (don't have one to test!)

### Key Vault

terraform doesn't support the "All" permission anymore but you may still have vaults using that permission.

### Virtual Network Gateway

if no bgp settings specified for VNet Gateway, `terraform plan` will report 
a benign change
 [see issue in github](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1993)

	~ update in-place
	Terraform will perform the following actions:

	~ azurerm_virtual_network_gateway.rg-$RGNAME__vgw-$VGWNAME
		bgp_settings.#: "" => <computed>


CHANGELOG.md
# az2tf (Python version) - changelog

## (December 2020)
* Added support for azure_app_service_slot

## (December 2019)
* Updates for Terraform 12 Interpolation and other depreciations

## (November 2019)
* Major change - now uses python3 ahead of depreciation of python2
* Support multiple clouds AzureCloud, AzureChinaCloud, AzureGermanCloud & AzureUSGovernment

## (August 10 2019)
* added zurerm_eventhub_namespace_authorization_rule
## (August 8 2019)
* added support for azure_eventhub, azurerm_eventhub_namespace 

## (June 21 2019)
* added support for azurerm_policy_set_definition

## (June 19 2019)
* added support for azurerm_shared_image_gallery
* added support for azurerm_shared_image
* added support for azurerm_shared_image_version

## (June 07, 2019)
* added support for azurerm_logic_app_workflow
* added support for azurerm_virtual_machine_extension

## (June 06, 2019)
* Completed port from bash version to this Python version 



auth1.py
#from azure.common.client_factory import get_client_from_cli_profile
#from azure.mgmt.compute import ComputeManagementClient
#from azure.mgmt.resource import ResourceManagementClient
import subprocess
import requests
import adal
import os
import json
import sys
import glob
import argparse

parser = argparse.ArgumentParser(description='terraform sub rg')
parser.add_argument('-s', help='Subscription Id')
parser.add_argument('-r', help='Resource Group')
args = parser.parse_args()
csub=args.s
crg=args.r

if csub is not None:
    print("sub=" + csub) 
    # validate sub
if crg is not None:
    print("resource group=" + crg)
    # validate rg


if sys.version_info[0] > 2:
    #raise Exception("Must be using Python 2")
    print("Python version ", sys.version_info[0], " version 2 required, Exiting")
    exit()

def printf(format, *values):
    print(format % values )

#with open(filename, 'w') as f:
    #print >> f, 'Filename:'


#tenant = os.environ['TENANT']
#authority_url = 'https://login.microsoftonline.com/' + tenant
#client_id = os.environ['CLIENTID']
#client_secret = os.environ['CLIENTSECRET']
#resource = 'https://management.azure.com/'
#context = adal.AuthenticationContext(authority_url)
#token = context.acquire_token_with_client_credentials(resource, client_id, client_secret)
#headers = {'Authorization': 'Bearer ' + token['accessToken'], 'Content-Type': 'application/json'}
#params = {'api-version': '2016-06-01'}
#url = 'https://management.azure.com/' + 'subscriptions'
#r = requests.get(url, headers=headers, params=params)
#print(json.dumps(r.json(), indent=4, separators=(',', ': ')))

p = subprocess.Popen('az account get-access-token -o json', shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
c=0
for line in p.stdout.readlines():
    if "accessToken" in line:
        tk=line.split(":")[1].strip(' ",')
        tk2=tk.replace(",", "")
        bt2=tk2.replace('"', '')
    if "subscription" in line:
        tk=line.split(":")[1].strip(' ",')
        tk2=tk.replace(",", "")
        sub2=tk2.replace('"', '')
retval = p.wait()
if csub is not None:
    sub=csub
else:
    sub=sub2.rstrip('\n')

bt=bt2.rstrip('\n')
print "Subscription:",sub
headers = {'Authorization': 'Bearer ' + bt, 'Content-Type': 'application/json'}


fresfilename="data.json"
fres=open(fresfilename, 'w')
url="https://management.azure.com/subscriptions/" + sub + "/resources"
params = {'api-version': '2018-11-01'}
try: 
    r = requests.get(url, headers=headers, params=params)
    res= r.json()["value"]
except KeyError:
    print "Error getting resources"
    exit()
fres.write(json.dumps(res, indent=4, separators=(',', ': ')))
fres.close()



frgfilename="azurerm_resource_group.json"
frg=open(frgfilename, 'w')
url="https://management.azure.com/subscriptions/" + sub + "/resourceGroups"

params = {'api-version': '2014-04-01'}
r = requests.get(url, headers=headers, params=params)
rgs= r.json()["value"]
frg.write(json.dumps(rgs, indent=4, separators=(',', ': ')))
frg.close()


fnsgfilename="azurerm_network_security_group.json"
fnsg=open(fnsgfilename, 'w')
url="https://management.azure.com/subscriptions/" + sub + "/providers/Microsoft.Network/networkSecurityGroups"
params = {'api-version': '2018-07-01'}
r = requests.get(url, headers=headers, params=params)
nsg= r.json()["value"]
#print (json.dumps(nsg, indent=4, separators=(',', ': ')))
fnsg.write(json.dumps(nsg, indent=4, separators=(',', ': ')))
fnsg.close()

rfilename="resources2.txt"
fr=open(rfilename, 'w')
nprfilename="noprovider2.txt"
np=open(nprfilename, 'w')


count=len(res)-1
print count
for j in range(0, count):
    t1=res[j]
    name=res[j]['name']
    id=res[j]['id']
    rg=id.split("/")[4]
    try:
        isext=id.split("/")[9]
    except IndexError:
        isext=""

    loc=res[j]['location']
    rtype=res[j]['type']
    #print rtype

    if rtype == "Microsoft.Compute/availabilitySets":
        prov="azurerm_availability_set"
        fr.write(rg + ":" + prov + "\n")
    elif rtype == "Microsoft.Network/networkSecurityGroups":
        prov="azurerm_network_security_group"
        fr.write(rg + ":" + prov + "\n")

    elif rtype == "Microsoft.Storage/storageAccounts": 
        prov="azurerm_storage_account"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_storage_share"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_storage_container"
        fr.write(rg + ":" + prov + "\n")

    elif rtype == "Microsoft.Network/networkSecurityGroups":
        prov="azurerm_network_security_group"
        fr.write(rg + ":" + prov + "\n")

    elif rtype == "Microsoft.Compute/virtualMachines": 
        #echo $isext
        if isext != "extensions":
            prov="azurerm_virtual_machine"
            fr.write(rg + ":" + prov + "\n")
                
    elif rtype == "Microsoft.Network/networkInterfaces": 
        prov="azurerm_network_interface"
        fr.write(rg + ":" + prov + "\n")
    
    elif rtype == "Microsoft.Compute/disks":
        prov="azurerm_managed_disk"
        fr.write(rg + ":" + prov + "\n")
        
    elif rtype == "Microsoft.Automation/automationAccounts": 
        prov="azurerm_automation_account"
        fr.write(rg + ":" + prov + "\n")
          
    elif rtype == "Microsoft.Network/virtualNetworks":
        prov="azurerm_virtual_network"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_subnet"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_virtual_network_peering"
        fr.write(rg + ":" + prov + "\n")

    elif rtype == "Microsoft.Network/publicIPAddresses":
        prov="azurerm_public_ip"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Network/loadBalancers":
        prov="azurerm_lb"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_lb_nat_rule"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_lb_nat_pool"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_lb_backend_address_pool"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_lb_probe"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_lb_rule"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Network/virtualNetworkGateways":
        prov="azurerm_virtual_network_gateway"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Network/connections":
        prov="azurerm_virtual_network_gateway_connection"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Network/routeTables": 
        prov="azurerm_route_table"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.OperationalInsights/workspaces":
        prov="azurerm_log_analytics_workspace"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype ==  "Microsoft.OperationsManagement/solutions":
        prov="azurerm_log_analytics_solution"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.KeyVault/vaults":
        prov="azurerm_key_vault"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_key_vault_secret"
        fr.write(rg + ":" + prov + "\n")

    elif rtype == "Microsoft.RecoveryServices/vaults":
        prov="azurerm_recovery_services_vault"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.ContainerRegistry/registries":
        prov="azurerm_container_registry"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.ContainerService/managedClusters":
        prov="azurerm_kubernetes_cluster"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Network/localNetworkGateways":
        prov="azurerm_local_network_gateway"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Network/expressRouteCircuits":
        prov="azurerm_express_route_circuit"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_express_route_circuit_authorization"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_express_route_circuit_peering"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Compute/images": 
        prov="azurerm_image"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Network/networkWatchers": 
        prov="azurerm_network_watcher"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Network/applicationSecurityGroups":
        prov="azurerm_application_security_group"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.ContainerInstance/containerGroups":
        prov="azurerm_container_group"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Network/applicationGateways": 
        prov="azurerm_application_gateway"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.DocumentDb/databaseAccounts":
        prov="azurerm_cosmosdb_account"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.ServiceBus/namespaces": 
        prov="azurerm_servicebus_namespace"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_servicebus_queue"
        fr.write(rg + ":" + prov + "\n")
                  
    elif rtype == "Microsoft.Network/trafficmanagerprofiles":
        prov="azurerm_traffic_manager_profile"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_traffic_manager_endpoint"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Web/serverFarms": 
        prov="azurerm_app_service_plan"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Web/sites": 
        prov="azurerm_app_service"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_function_app"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Compute/virtualMachineScaleSets":
        prov="azurerm_virtual_machine_scale_set"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.ManagedIdentity/userAssignedIdentities":
        prov="azurerm_user_assigned_identity"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Compute/snapshots":
        prov="azurerm_snapshot"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Databricks/workspaces":
        prov="azurerm_databricks_workspace"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Sql/servers": 
        prov="azurerm_sql_server"
        fr.write(rg + ":" + prov + "\n")
        prov="azurerm_sql_database"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype == "Microsoft.Network/dnszones": 
        prov="azurerm_dns_zone"
        fr.write(rg + ":" + prov + "\n")
            
    elif rtype ==  "microsoft.insights/autoscalesettings":
        prov="azurerm_monitor_autoscale_setting"
        fr.write(rg + ":" + prov + "\n")
            
    else:
        np.write(rtype + "\n")

fr.close()
np.close()

# sort unique and fileter for Resource Group
rfilename="resources.txt"
fr=open(rfilename, 'w')
with open('resources2.txt', 'r') as r:
    for line in sorted(set(r)):
        trg=line.split(":")[0]
        if crg is not None:   # Resource Group Filter
            if trg == crg :
                fr.write(line,)
        else:
            fr.write(line,)
r.close()
fr.close()

# sort unique and fileter for Resource Group
rfilename="noprovider.txt"
fr=open(rfilename, 'w')
with open('noprovider2.txt', 'r') as r:
    for line in sorted(set(r)):
        fr.write(line,)

r.close()
fr.close()
os.remove('tf-staterm.sh')
os.remove('tf-stateimp.sh')
tfrm=open("tf-staterm.sh", 'a')
tfim=open("tf-stateimp.sh", 'a')

# handle resource groups
tfp="azurerm_resource_group"
tffile=tfp+"*.tf"
fileList = glob.glob(tffile) 
# Iterate over the list of filepaths & remove each file.
for filePath in fileList:
    try:
        os.remove(filePath)
    except:
        print("Error while deleting file : ", filePath)

count=len(rgs)-1
print count
for j in range(0, count):
    
    name=rgs[j]["name"]
    
    if crg is not None:
        if name != crg:
            continue
    
    rname=name.replace(".","-")
    prefix=tfp+"__"+rname
    
    rfilename=prefix+".tf"
    fr=open(rfilename, 'w')
    fr.write("")
    fr.write('resource ' + tfp + ' ' + rname + ' {\n')
    fr.write('\t name = "' + name + '"\n')
    fr.write('\t location = "'+ loc + '"\n')
    fr.write('}\n')  

# tags block
    try:
        mtags=rgs[j]["tags"]
    except:
        mtags="{}"
    tcount=len(mtags)-1
    if tcount > 1 :
        fr.write('tags { \n')
        print tcount
        for key in mtags.keys():
            tval=mtags[key]
            fr.write('\t "' + key + '"="' + tval + '"\n')
        #print(json.dumps(rgs[j], indent=4, separators=(',', ': ')))
        print(json.dumps(mtags, indent=4, separators=(',', ': ')))
        fr.write('} \n')
    fr.close()
    tfrm.write('terraform state rm '+tfp+'.'+rname + '\n')
    tfim.write('terraform state import '+tfp+'.'+rname+' '+id+ '\n')

tfrm.close()
tfim.close()
exit()

rclient = get_client_from_cli_profile(ResourceManagementClient)

for resource_group in rclient.resource_groups.list():
    print(resource_group.name)
